neo/idlib/math/Simd_AVX.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+===========================================================================
+
+Doom 3 GPL Source Code
+Copyright (C) 1999-2011 id Software LLC, a ZeniMax Media company.
+
+This file is part of the Doom 3 GPL Source Code ("Doom 3 Source Code").
+
+Doom 3 Source Code is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Doom 3 Source Code is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Doom 3 Source Code.  If not, see <http://www.gnu.org/licenses/>.
+
+In addition, the Doom 3 Source Code is also subject to certain additional terms. You should have received a copy of these additional terms immediately following the terms and conditions of the GNU General Public License which accompanied the Doom 3 Source Code.  If not, please request a copy in writing from id Software at the address below.
+
+If you have questions concerning this license or the applicable additional terms, you may contact in writing id Software LLC, c/o ZeniMax Media Inc., Suite 120, Rockville, Maryland 20850 USA.
+
+===========================================================================
+*/
+
+#include "sys/platform.h"
+#include "Simd_AVX.h"
+
+//===============================================================
+//
+//  AVX implementation of idSIMDProcessor
+//
+//===============================================================
+
+// Revelator: these work whether in gcc clang or msvc in x86 or x64 (no inline assembly used)
+#if defined(_MSC_VER) && ( defined(_M_X64) || defined(_M_IX86) ) || \
+    defined(__GNUC__) && ( defined(__i386__) || defined (__x86_64__) ) && defined(__AVX__)
```

Comment:
With GCC, this currently disabled the code because __AVX__ is only defined if `-mavx` is used - which would have to be set in CMakeLists.txt, only for this file

Reply:
also, I guess `__AVX__` is only set if `__i386__` or `__x86_64__` is set, so those checks aren't necessary

Reply:
yeah i been working on some cmake changes to probe the users cpu and set the correct flags for gnu, still not ready but ill post the change when it is.

but is it only avx that needs a flag set or do we also need it for every SSE type ? like -sse -sse2 -sse3 -ssse3 -mavx -mavx2 etc. if so that would suck.

Reply:
hmm maybe this could also be handled in the source files instead of CMake, like
```c++
// add this at the beginning of the source file, under the big comment block with the license
#ifdef __GNUC__
#pragma GCC target (avx)
#endif
```

https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html#index-target-function-attribute-5 has a list of valid target arguments

Reply:
apparently for clang it's slighly more complicated, see https://stackoverflow.com/a/56395854

so maybe it must be something like
```c++
#ifdef __clang__
#pragma clang attribute push (__attribute__((target("avx"))), apply_to=function)
#elif defined __GNUC__
#pragma GCC target ("avx")
#endif
```

though it's possible that in practice the older extensions (>SSE2) must also be listed, if the code uses intrinsics from them as well, that might look like
`#pragma clang attribute push (__attribute__((target("sse3", "ssse3", "avx"))), apply_to=function)` or `#pragma GCC target ("sse3", "ssse3", "avx")` (just an example if SSE3, SSSE3 and AVX are needed)

Reply:
hmm interresting i had no idea gcc supported setting intrinsics via pragmas but cool if it does :) 
guess we need to ask around the clang user base for specifics luckily i know one of the devs martin storsjoe he might have some insight on this.

Reply:
hmm yes for clang this does seem the right approach allthough we also need to pop the stack afterwards.
for gcc #pragma GCC target("avx2,fma") should also work for avx2. can even set tune to the lowest cpu supporting it for more fine control like eg. tune=haswell maybe ?. also seems the pragmas need to come after the intrinsics header otherwise they break.

darkmod has made a lot of changes to intrinsics for one they ditched all the older simd code using the portable intrinsics instead.

Reply:
```
//===============================================================
//
//  AVX2 implementation of idSIMDProcessor
//
//===============================================================

// Revelator: these work whether in gcc clang or msvc in x86 or x64 (no inline assembly used)
#include <immintrin.h>

// must come after immintrin.h
#if defined(__i386__) || defined (__x86_64__)
#ifdef __clang__
#pragma clang attribute push (__attribute__((target("avx2,fma"))), apply_to=function)
#elif defined __GNUC__
#pragma GCC target ("avx2,fma")
#endif
#endif // x86 or x86_64

#include "idlib/geometry/DrawVert.h"
#include "idlib/geometry/JointTransform.h"
#include "idlib/math/Vector.h"
#include "idlib/math/Plane.h"

/*
============
idSIMD_AVX2::GetName
============
*/
const char *idSIMD_AVX2::GetName( void ) const {
    return "MMX & SSE & SSE2 & SSE3 & AVX & AVX2";
}

/*
============
idSIMD_AVX2::CullByFrustum
============
*/
void VPCALL idSIMD_AVX2::CullByFrustum( idDrawVert *verts, const int numVerts, const idPlane frustum[6], byte *pointCull, float epsilon ) {
    const __m256 fA = _mm256_set_ps( 0, 0, frustum[5][0], frustum[4][0], frustum[3][0], frustum[2][0], frustum[1][0], frustum[0][0] );
    const __m256 fB = _mm256_set_ps( 0, 0, frustum[5][1], frustum[4][1], frustum[3][1], frustum[2][1], frustum[1][1], frustum[0][1] );
    const __m256 fC = _mm256_set_ps( 0, 0, frustum[5][2], frustum[4][2], frustum[3][2], frustum[2][2], frustum[1][2], frustum[0][2] );
    const __m256 fD = _mm256_set_ps( 0, 0, frustum[5][3], frustum[4][3], frustum[3][3], frustum[2][3], frustum[1][3], frustum[0][3] );
    const __m256 eps = _mm256_set1_ps( epsilon );
    const byte mask6 = ( 1 << 6 ) - 1;

    for ( int j = 0; j < numVerts; j++ ) {
        idVec3 &vec = verts[j].xyz;
        __m256 vX = _mm256_set1_ps( vec.x );
        __m256 vY = _mm256_set1_ps( vec.y );
        __m256 vZ = _mm256_set1_ps( vec.z );
        __m256 d =  _mm256_fmadd_ps( fA, vX,
                    _mm256_fmadd_ps( fB, vY,
                    _mm256_fmadd_ps( fC, vZ, fD ) ) );
        int mask_lo = _mm256_movemask_ps( _mm256_cmp_ps( d, eps, _CMP_LT_OQ ) );
        pointCull[j] = ( byte )( mask_lo & mask6 ); // gcc compiler warning
    }
    _mm256_zeroupper();
}

/*
============
idSIMD_AVX2::CullByFrustum2
============
*/
void VPCALL idSIMD_AVX2::CullByFrustum2( idDrawVert *verts, const int numVerts, const idPlane frustum[6], unsigned short *pointCull, float epsilon ) {
    const __m256 fA = _mm256_set_ps( 0, 0, frustum[5][0], frustum[4][0], frustum[3][0], frustum[2][0], frustum[1][0], frustum[0][0] );
    const __m256 fB = _mm256_set_ps( 0, 0, frustum[5][1], frustum[4][1], frustum[3][1], frustum[2][1], frustum[1][1], frustum[0][1] );
    const __m256 fC = _mm256_set_ps( 0, 0, frustum[5][2], frustum[4][2], frustum[3][2], frustum[2][2], frustum[1][2], frustum[0][2] );
    const __m256 fD = _mm256_set_ps( 0, 0, frustum[5][3], frustum[4][3], frustum[3][3], frustum[2][3], frustum[1][3], frustum[0][3] );
    const __m256 eps = _mm256_set1_ps( epsilon );
    static const __m256 epsM = _mm256_set1_ps( -epsilon );
    const short mask6 = ( 1 << 6 ) - 1;

    for ( int j = 0; j < numVerts; j++ ) {
        idVec3 &vec = verts[j].xyz;
        __m256 vX = _mm256_set1_ps( vec.x );
        __m256 vY = _mm256_set1_ps( vec.y );
        __m256 vZ = _mm256_set1_ps( vec.z );
        __m256 d =  _mm256_fmadd_ps( fA, vX,
                    _mm256_fmadd_ps( fB, vY,
                    _mm256_fmadd_ps( fC, vZ, fD ) ) );
        int mask_lo = _mm256_movemask_ps( _mm256_cmp_ps( d, eps, _CMP_LT_OQ ) );
        int mask_hi = _mm256_movemask_ps( _mm256_cmp_ps( d, eps, _CMP_GT_OQ ) );
        pointCull[j] = ( unsigned short )( ( mask_lo & mask6 ) | ( mask_hi & mask6 ) ) << 6; // gcc compiler warning
    }
    _mm256_zeroupper();
}

#if defined(__i386__) || defined (__x86_64__)
#ifdef __clang__
#pragma clang attribute pop
#endif
#endif
```

something like this maybe ?

Reply:
seems we also need #pragma GCC push_options before the #pragma GCC target string and subsequent stack pop as well for gcc.

Reply:
```
#if defined(__i386__) || defined (__x86_64__)
#ifdef __clang__
#pragma clang attribute push (__attribute__((target("avx,avx2,fma,arch=haswell"))), apply_to=function)
#elif defined __GNUC__
#pragma GCC push_options
#pragma GCC target ("avx,avx2,fma,arch=haswell")
#endif
#endif // x86 or x86_64
```
there we go avx2 depends on avx and fma but as far as i can see not on any of the older sse mechanics.

Reply:
could be, TBH I wasn't even sure if this works at all, was just an idea because I know you can locally enable/disable optimization like this

my assumption about pushing/popping was that it doesn't matter because we want those options to be used for the whole source file - but could be that compilers insist on it anyway, IDK

Reply:
BTW: `// must come after immintrin.h` why? I'd think that the intrinsics header also needs to know we're now targeting AVX instead of just SSE2

Reply:
Another thing: I don't think `arch=haswell` should be set.
We don't want the generated code to be optimized for haswell, or to even use instructions that are only available on haswell but not some other CPUs that might support avx2 (or whatever) anyway (for example by using instructions from other CPU extensions that only Intel supports, but AMD doesn't).

The GCC manpage says:
>   `-march=cpu-type`
>       Generate instructions for the machine type cpu-type.  In contrast to `-mtune=cpu-type`, which merely tunes the generated code for the specified cpu-type, `-march=cpu-type` allows GCC to **generate code that may _not run at all_ on processors other than the one indicated**.  Specifying `-march=cpu-type` implies `-mtune=cpu-type`, except where noted otherwise.

Reply:
oh was not ment for haswell optimization just to make sure we dont use the code on a cpu that does not support it, otherwise it will surely crash. or should we use -mtune ?.

the other idea sounds great :)

Reply:
oh fotgot the reason it must come after the intrinsic header is because otherwise the pragma fails, this was noticed by someone doing something similar so i took my cue from that :)

Reply:
>  just to make sure we dont use the code on a cpu that does not support it, otherwise it will surely crash. 

yes, and that's exactly why -march should **not** be set :-p

> or should we use -mtune ?.

no, just use "avx2,fma" if the intrinsics code in that file needs AVX2 and FMA

Then the compiler should only use instructions that are either available in x86 with SSE and SSE2, **or** the extensions specified in the pragma (AVX2 and FMA in this case).

Note: x86_64 aka x64 always supports SSE and SSE2, and when building 32bit x86 code with MSVC it nowadays also defaults to supporting SSE2 (though that's not relevant for these GCC/clang-specific macros). Currently when using GCC/clang to build for 32bit x86 dhewm3 does *not* use SSE(2), but I plan to change that in CMake.

Reply:
> oh forgot the reason it must come after the intrinsic header is because otherwise the pragma fails, this was noticed by someone doing something similar

We should try this ourselves to see if that's actually (still) true.

Reply:
im game :)

Reply:
avx2 needs avx as well according to darkmod so avx,avx2,fma it seems.

Reply:
that sounds plausible, use that then :)

(no idea if `-mavx2` implies `-mavx` - it would make sense, I don't think there are any CPUs implementing AVX2 but not AVX, but IDK, the manpage doesn't mention it)

Reply:
btw just for giggles and to see if my sdl wrappers in cpu.cpp worked i tried a build with SDL1.5 and oh boy...
needs some missing gl definitions not supplied by SDL1 and there was this one in SDL_win32_main.c

```
	/* Load SDL dynamic link library */
	if ( SDL_Init(SDL_INIT_NOPARACHUTE) < 0 ) {
		ShowError("WinMain() error", SDL_GetError());
		return(FALSE);
	}
	//atexit(cleanup_output); // uh referencing a static function in another source file ?
	atexit(cleanup);
```
note the one i commented.

so after pulling in all the gl definitions and commenting that one it runs but the mouse is all wonky (sometimes dissapearing) possible oversight ? or should that function not have been static maybe ?.

Reply:
ok ill go with darkmods suggestion :) shouldnt hurt at least.

Reply:
agree i just thought you keept the SDL1 code for posterity or possibly some ancient machine ;)
guess it should be removed at some point.

hmm SDL_win32_main.cpp also only seems to be pulled in if you specifically request SDL1.

Reply:
i foresee one problem atleast with my hybrid glsl/arb renderer since SDL3 no longer supports the old hardware gamma we would need shader code for the glsl interactions for gamma as well when moving on to SDL3.

but lets get the intrinsics working first :)

Reply:
> hmm SDL_win32_main.cpp also only seems to be pulled in if you specifically request SDL1.

oh right, now I remember.
yes, that `atexit(cleanup_output);` line can be deleted, `cleanup_output()` has been moved to win_main.cpp and `atexit(cleanup_output);` is now called from `SDL_main()` in win_main.cpp (for all SDL versions)

Reply:
I'm not sure I want GLSL at all at this point, but if it's added it definitely needs to support shader gamma.

Reply:
normally would not be a problem but because of the way it is loaded directly in RB_ARB2_DrawInteraction it causes problems for my renderer because it does the housekeeping itself if told this one goes to arb then it goes to the arb side and vice versa but since it is loaded directly in the interaction renderer it bonks out if glsl is selected.
as an example i can run sikkmod with glsl interactions and still use all the arb effects :)